s2a/src/main/java/io/grpc/s2a/S2AChannelCredentials.java

```diff
@@ -0,0 +1,130 @@
+  public static Builder createBuilder(String s2aAddress) {
+    checkArgument(!isNullOrEmpty(s2aAddress), "S2A address must not be null or empty.");
+    return new Builder(s2aAddress);
+  }
+
+  /** Builds an {@code S2AChannelCredentials} instance. */
+  @NotThreadSafe
+  public static final class Builder {
+    private final String s2aAddress;
+    private ObjectPool<Channel> s2aChannelPool;
+    private Optional<ChannelCredentials> s2aChannelCredentials;
+    private @Nullable S2AIdentity localIdentity = null;
+
+    Builder(String s2aAddress) {
+      this.s2aAddress = s2aAddress;
+      this.s2aChannelPool = null;
+      this.s2aChannelCredentials = Optional.empty();
+    }
+
+    /**
+     * Sets the local identity of the client in the form of a SPIFFE ID. The client may set at most
+     * 1 local identity. If no local identity is specified, then the S2A chooses a default local
+     * identity, if one exists.
+     */
+    @CanIgnoreReturnValue
+    public Builder setLocalSpiffeId(String localSpiffeId) {
+      checkNotNull(localSpiffeId);
+      checkArgument(localIdentity == null, "localIdentity is already set.");
+      localIdentity = S2AIdentity.fromSpiffeId(localSpiffeId);
+      return this;
+    }
+
+    /**
+     * Sets the local identity of the client in the form of a hostname. The client may set at most 1
+     * local identity. If no local identity is specified, then the S2A chooses a default local
+     * identity, if one exists.
+     */
+    @CanIgnoreReturnValue
+    public Builder setLocalHostname(String localHostname) {
+      checkNotNull(localHostname);
+      checkArgument(localIdentity == null, "localIdentity is already set.");
+      localIdentity = S2AIdentity.fromHostname(localHostname);
+      return this;
+    }
+
+    /**
+     * Sets the local identity of the client in the form of a UID. The client may set at most 1
+     * local identity. If no local identity is specified, then the S2A chooses a default local
+     * identity, if one exists.
+     */
+    @CanIgnoreReturnValue
+    public Builder setLocalUid(String localUid) {
+      checkNotNull(localUid);
+      checkArgument(localIdentity == null, "localIdentity is already set.");
+      localIdentity = S2AIdentity.fromUid(localUid);
+      return this;
+    }
+
+    /** Sets the credentials to be used when connecting to the S2A. */
+    @CanIgnoreReturnValue
+    public Builder setS2AChannelCredentials(ChannelCredentials s2aChannelCredentials) {
+      this.s2aChannelCredentials = Optional.of(s2aChannelCredentials);
+      return this;
+    }
+
+    public ChannelCredentials build() {
+      checkState(!isNullOrEmpty(s2aAddress), "S2A address must not be null or empty.");
+      ObjectPool<Channel> s2aChannelPool =
+          SharedResourcePool.forResource(
+              S2AHandshakerServiceChannel.getChannelResource(s2aAddress, s2aChannelCredentials));
+      checkNotNull(s2aChannelPool, "s2aChannelPool");
+      this.s2aChannelPool = s2aChannelPool;
```

Comment:
Try to avoid mutating builders in the `build()` method. There seems to be no need to use a field here. You can just pass this pool to `buildProtocolNegotiatorFactory()` or (better) have `buildProtocolNegotiatorFactory()` construct the `s2aChannelPool` itself. Or you inline `buildProtocolNegotiatorFactory()` into the `build()` method.

IIRC, ALTS had `buildProtocolNegotiatorFactory()` for use in the older GoogleDefaultChannelBuilder API and related. In your case, you don't need that, so you probably don't need the method split out.

Reply:
Thanks for pointing that out, https://github.com/grpc/grpc-java/pull/11573 constructs the ```s2aChannelPool``` in ```buildProtocolNegotiatorFactory```